hibernate-core/src/main/java/org/hibernate/loader/plan/build/internal/FetchStyleLoadPlanBuildingAssociationVisitationStrategy.java

```diff
@@ -190,11 +200,107 @@ protected FetchStrategy adjustJoinFetchIfNeeded(
+						if ( elementIdTypeEmbedded.getSubtypes().length == 1 &&
+								elementIdTypeEmbedded.getPropertyNames()[ 0 ].equals( collectionPersister.getMappedByProperty() ) ) {
+							// The associated entity's ID is the other (many-to-one) side of the association.
+							// The one-to-many side must be set to FetchMode.SELECT; otherwise,
+							// there will be an infinite loop because the current entity
+							// would need to be loaded before the associated entity can be loaded,
+							// but the associated entity cannot be loaded until after the current
+							// entity is loaded (since the current entity is the associated entity's ID).
+							return new FetchStrategy( fetchStrategy.getTiming(), FetchStyle.SELECT );
+						}
+					}
+				}
+			}
 		}
 
+		if ( attributeType.isEntityType() &&
+				fetchStrategy.getTiming() == FetchTiming.IMMEDIATE &&
+				fetchStrategy.getStyle() == FetchStyle.JOIN ) {
+			final EntityType entityType = (EntityType) attributeType;
+			final EntityReference currentEntityReference = currentSource.resolveEntityReference();
+			if ( currentEntityReference != null ) {
+				final EntityPersister currentEntityPersister = currentEntityReference.getEntityPersister();
+				if ( entityType.isOneToOne() && entityType.getForeignKeyDirection() == ForeignKeyDirection.TO_PARENT ) {
+					// attributeDefinition is the "mappedBy" (inverse) side of a
+					// bidirectional one-to-one association.
+					final OneToOneType oneToOneType = (OneToOneType) attributeType;
+					final String associatedUniqueKeyPropertyName = oneToOneType.getIdentifierOrUniqueKeyPropertyName(
+							sessionFactory()
+					);
+					if ( associatedUniqueKeyPropertyName == null ) {
+						// The foreign key for the other side of the association is the ID.
+						// Now check if the ID itself is the other side of the association.
+						final EntityPersister associatedEntityPersister = (EntityPersister) oneToOneType.getAssociatedJoinable(
+								sessionFactory()
+						);
+						final Type associatedIdentifierType = associatedEntityPersister.getIdentifierType();
+						if ( associatedIdentifierType.isComponentType() &&
+								( (CompositeType) associatedIdentifierType ).isEmbedded() ) {
+							final EmbeddedComponentType associatedNonEncapsulatedIdentifierType =
+									(EmbeddedComponentType) associatedIdentifierType;
+							if ( associatedNonEncapsulatedIdentifierType.getSubtypes().length == 1 &&
+									EntityType.class.isInstance( associatedNonEncapsulatedIdentifierType.getSubtypes()[0] ) ) {
+								final EntityType otherSideEntityType =
+										( (EntityType) associatedNonEncapsulatedIdentifierType.getSubtypes()[0] );
+								if ( otherSideEntityType.isLogicalOneToOne() &&
+										otherSideEntityType.isReferenceToPrimaryKey() &&
+										otherSideEntityType.getAssociatedEntityName().equals( currentEntityPersister.getEntityName() ) ) {
+									// The associated entity's ID is the other side of the association.
```

Comment:
Similar thought about code comment here...

```
// The owning side of the inverse to-one is defined by the associated entity's id.
// 
// Because of how Loaders process ids when processing the ResultSet, this condition would 
// lead to an infinite loop.  Adjust the fetch to use a SELECT fetch instead of JOIN to avoid
```